File: frontend/src/services/notifications.ts
import { getToken } from './auth';

const API_BASE_URL = import.meta.env.VITE_API_URL || '';

function getAuthHeaders() {
  const token = getToken();
  return {
    'Authorization': `Bearer ${token}`,
    'Content-Type': 'application/json',
  };
}

export interface NotificationPreferences {
  email_notifications: boolean;
  webhook_url?: string;
  event_notifications: boolean;
  camera_status_notifications: boolean;
  system_alerts: boolean;
}

export interface NotificationPreferencesUpdate {
  email_notifications?: boolean;
  webhook_url?: string;
  event_notifications?: boolean;
  camera_status_notifications?: boolean;
  system_alerts?: boolean;
}

export interface NotificationStatus {
  websocket_connected: boolean;
  email_enabled: boolean;
  webhook_configured: boolean;
  active_connections: number;
}

export interface TestEmailResponse {
  message: string;
}

// Get notification preferences
export async function getNotificationPreferences(): Promise<NotificationPreferences> {
  const response = await fetch(`${API_BASE_URL}/api/v1/notifications/preferences`, {
    headers: getAuthHeaders(),
  });
  
  if (!response.ok) {
    throw new Error(`Failed to get notification preferences: ${response.statusText}`);
  }
  
  return response.json();
}

// Update notification preferences
export async function updateNotificationPreferences(
  preferences: NotificationPreferencesUpdate
): Promise<NotificationPreferences> {
  const response = await fetch(`${API_BASE_URL}/api/v1/notifications/preferences`, {
    method: 'PUT',
    headers: {
      ...getAuthHeaders(),
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(preferences),
  });
  
  if (!response.ok) {
    throw new Error(`Failed to update notification preferences: ${response.statusText}`);
  }
  
  return response.json();
}

// Get notification status
export async function getNotificationStatus(): Promise<NotificationStatus> {
  const response = await fetch(`${API_BASE_URL}/api/v1/notifications/status`, {
    headers: getAuthHeaders(),
  });
  
  if (!response.ok) {
    throw new Error(`Failed to get notification status: ${response.statusText}`);
  }
  
  return response.json();
}

// Send test email
export async function sendTestEmail(): Promise<TestEmailResponse> {
  const response = await fetch(`${API_BASE_URL}/api/v1/notifications/test-email`, {
    method: 'POST',
    headers: getAuthHeaders(),
  });
  
  if (!response.ok) {
    throw new Error(`Failed to send test email: ${response.statusText}`);
  }
  
  return response.json();
} 

File: backend/app/schemas/notifications.py
from pydantic import BaseModel, EmailStr, HttpUrl
from typing import Optional, Dict, Any
from datetime import datetime

class NotificationPreferences(BaseModel):
    """Notification preferences for a user"""
    email_notifications: bool = True
    webhook_url: Optional[str] = None
    event_notifications: bool = True
    camera_status_notifications: bool = True
    system_alerts: bool = True

class NotificationPreferencesUpdate(BaseModel):
    """Update notification preferences"""
    email_notifications: Optional[bool] = None
    webhook_url: Optional[str] = None
    event_notifications: Optional[bool] = None
    camera_status_notifications: Optional[bool] = None
    system_alerts: Optional[bool] = None

class NotificationMessage(BaseModel):
    """Real-time notification message"""
    type: str
    title: str
    message: str
    data: Dict[str, Any]
    timestamp: str
    priority: str = "normal"
    category: str = "general"

class NotificationStatus(BaseModel):
    """Notification system status"""
    websocket_connected: bool
    email_enabled: bool
    webhook_configured: bool
    active_connections: int

class TestEmailResponse(BaseModel):
    """Response for test email endpoint"""
    message: str 